### ISSUE TITLE
Saved mp3 metadata with album art do not show the album art as icon in Win10

### ISSUE THREAD

**Issue by ElmoFuntz** [Zeugma440/atldotnet#48]
## The problem

Loaded an MP3 with your test code for loading. Track info and image show correctly. Used your test code to save 2 lines of metadata (used verbatim). Image that was previously showing before in Windows 10 large icon view now shows the default mp3 icon. Note: the album art image was not one of the items changed.

Loaded the mp3 into mp3tag did nothing other than save the tag again and it nows shows again in windows.

No errors in log

## Environment

2.16 on Windows 10 1809.

Interestingly m4b files when saved do continue to correctly show their artwork.

**Comment by Zeugma440** [comment]
Well, that's a very good question.

The tests I made were all done with mp3tag, which should always be able to read the embedded pictures after they've been updated. I never checked if Windows could still see them.

I'm gonna do some tests and keep you updated.

**Comment by ElmoFuntz** [comment]
mp3tag did see the artwork for me as well.

**Comment by Zeugma440** [comment]
__TL;DR :__ Fixed for next release

__Detailed explanation__
ID3v2.4 specs state that the `APIC` (embedded picture) frame should start with `Text encoding      $xx`

According to that same spec, the "text encoding description byte" should be valued according to the encoding chosen to write strings. As ATL writes its ID3v2.4 tags using UTF-8 by default, that byte is written with the corresponding value, i.e. ` $03`

Now according to my tests, Windows seems to ignore the `APIC` frame entirely if it doesn't find `$00` (the code for the `ISO-8859-1` encoding, aka Latin-1, which covers western characters only) at that position. 

Forcing that value to `$00`, would make it impossible to write any picture description using an encoding different than `ISO-8859-1`. That being said, usage of the "picture description" field of ID3v2 is quite rare

=> being "Windows-friendly" by default won't hurt most users of the library. For those who _really_ want to write picture descriptions in japanese or cyrillic, a new `ID3v2_forceAPICEncodingToLatin1` option has been introduced in the library settings.

**Comment by ElmoFuntz** [comment]
Tested this with the mp3 format and wondering how the picture should work. At first I used your example code which had a PIC_TYPE of CD. This saved, did not crash, but also did not display anything in windows. I changed the type to Generic loaded the file back up and I can see the image was embedded because it populated my picturebox. I resaved and it does show artwork now in windows. Is this expected? Below is my code if it matters. I had to change your example since i'm reading from a picture box instead of disk.

```
            PictureInfo newPicture = new PictureInfo(Commons.ImageFormat.Jpeg, PictureInfo.PIC_TYPE.Generic);
            ImageConverter converter = new ImageConverter();
            newPicture.PictureData = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
            theTrack.EmbeddedPictures.Add(newPicture);
```

I'm also curious does the file specifications require you to specify the imageformat? I groaned a little at having to make a case or if with a variation of new PictureInfo for each format :)

**Comment by ElmoFuntz** [comment]
Ok so I started messing around loading images of type png even though I had jpeg selected to see how it behaved and I am seeing some issues I suspect with how its stored. Also oddly the "other" format is no longer showing as an icon. I changed it to PIC_TYPE front and its showing again.

Annnd now it's not showing again... I was all in the middle of typing up a bunch of bullets on what I am seeing but its weirdly inconsistent.  It also appears as if once you save a picture it saves the next one in the next position and never goes back to 1 unless you remove all the images. 

I'm trying to figure out how to check if there is an embedded image in the first position (or any after that) but so far striking out. I might have to go look at your code and see if you have an example.

Oh btw even with PIC_TYPE as png and using a jpeg it did embed the image (in that mp3tag saw it and displayed it) but all of my testing for the above was with it set for the correct type and on a clean file.

The above embed image also does not show in groove or WMP but it does in VLC.... The good news is it does not appear to corrupt the file for playback

**Comment by Zeugma440** [comment]
`I'm also curious does the file specifications require you to specify the imageformat? I groaned a little at having to make a case or if with a variation of new PictureInfo for each format :)`

ID3v2 and APE need to know the format of the picture because they write data about it.

That being said, I should be able to simplify the code user-side and auto-detect the image format according to its binary data instead of having the user specify it. I'll do some tests and keep you informed.

***

For the rest, I'm super sorry but what you've written is very confusing to me. I can't make out actual issues from ponderings and notes about testing your own software.

Would it be possible to post repro steps and/or sample code for any remaining issue ?

e.g.
```
1. Run the sample code
2. View the file on Windows

Expected : The cover is displayed as icon
Actual : No cover is displayed as icon
```

Thanks in advance 😉

**Comment by ElmoFuntz** [comment]
Yeah sorry it was getting late and I think I rambled... I was actually in the middle of doing exactly what you listed then things started acting inconsistant/strange and so did I :) Let me post something up now. I also want to double check and be sure i'm really using the right dll.

**Comment by ElmoFuntz** [comment]
I did some more testing and hopefully the below is not too rambling this time. Part of why I was a bit confusing above is that it appears the save is working fine with some test files and is not with others. I am going to send you the 2 files I am working with. works.mp3 updates every time and doesnt.mp3 does the below.

Steps taken
Load a test mp3 with embedded art
Load new art into picturebox
saved mp3

Results:
Windows art remains the same
Reload file in my program, new art is displayed
Copy file to new location in windows, old art is displayed
Copy file to new PC, old art is displayed
Load file in mp3tag, new art is displayed but so is the old art. There are 2 images embedded.

Next steps taken
Removed both images with mp3tag
Windows now shows standard mp3 icon
close mp3tag to avoid file locks
Load new art into my program
Save mp3

Results:
No art on desktop
Art shows in mp3tag
Art shows in my program

Save mp3 in mp3tag without making any changes and the art now shows on desktop as well.

After updating with mp3tag I am able to update the file one time and have it update in windows subsequent updates repeat the above steps.

The same type of issues as above also happen with images that do not already have embedded art.

My code below

```
        private void displayProgress(float progress)
        {
            //Console.WriteLine(progress + "%");
            txtProgress.Text = progress + "%";
        }

         private void btnSave_Click(object sender, EventArgs e)
        {
            // Adapt this to whatever your file path needs to be
            string fileName = "d:/test.mp3";

            IProgress<float> progress = new Progress<float>(displayProgress);
            // Load audio file information into memory
            Track theTrack = new Track(fileName, progress);

            // Modify metadata
            theTrack.Artist = "Hey ho";
            theTrack.Composer = "Oscar Wilde";
            theTrack.Album = "Fake album starts here and is longer than the original one";

            PictureInfo newPicture = new PictureInfo(Commons.ImageFormat.Jpeg, PictureInfo.PIC_TYPE.Front);
            ImageConverter converter = new ImageConverter();
            newPicture.PictureData = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
            theTrack.EmbeddedPictures.Add(newPicture);


            // Save modifications on the disc

            theTrack.Save();
        }
```

**Comment by Zeugma440** [comment]
This one is still a mystery to me, despite the clearer instructions you just gave (thanks for these, btw)

I have asked you for an additional resource (check your e-mail 😄 )

***

Meanwhile, there's something you should know : using the following code repeatedly on the same file adds one additional image with every iteration.

```csharp
new Track();
theTrack.EmbeddedPictures.Add(newPicture); // Adds a picture (not replace, but _add_)
theTrack.Save();
```csharp

If you want to replace the existing pictures, you should use `theTrack.EmbeddedPictures.Clear` or `theTrack.EmbeddedPictures.RemoveAt` before calling `theTrack.EmbeddedPictures.Add`

**Comment by ElmoFuntz** [comment]
I know about removeat but what I don't know is how to verify if a picture
exists before trying to remove it. So for now i'm commenting out the line
if one doesn't in order to not crash. I did not know about clear.

I was going to ask as well if there is a special way i need to remove
chapters once that gets going. Both all and single/multiples.

**Comment by Zeugma440** [comment]
```I don't know is how to verify if a picture exists before trying to remove it.```

Umm... how about adding `if (theTrack.EmbeddedPictures.Count > 0) theTrack.EmbeddedPictures.Clear()` just before `theTrack.EmbeddedPictures.Add` ?

***

`is there a special way to remove chapters once that gets going. Both all and single/multiples.`

Same convention : either use `theTrack.Chapters.Clear` or `theTrack.Chapters.RemoveAt`

**Comment by ElmoFuntz** [comment]
On Sun, Jan 26, 2020 at 7:18 AM Zeugma440 <notifications@github.com> wrote:

> I don't know is how to verify if a picture exists before trying to remove
> it.
>
> Umm... how about adding if (theTrack.EmbeddedPictures.Count > 0)
> theTrack.EmbeddedPictures.Clear() just before
> theTrack.EmbeddedPictures.Add ?
>

Lol I thought I tried count. I must have done something wrong. Thanks.


------------------------------

**Comment by Zeugma440** [comment]
I still don't get how Windows works. On my side, I have Windows 10 1903 and it detects everything I feed him through the library.

The file called "doesnt" does _not_ work either on my side but I don't understand why :
- ID3v2 picture frame looks good
- binary data of the embedded image looks good too

I'm starting to think Windows has an issue with ID3v2.4.

***

Here's the new version of the library.

[ATL.zip](https://github.com/Zeugma440/atldotnet/files/4113833/ATL.zip)

Please note that I've rewritten all constructors for `PictureInfo` so that you don't have to provide the picture format everytime.

=> You should replace

```csharp
PictureInfo newPicture = new PictureInfo(Commons.ImageFormat.Jpeg, PictureInfo.PIC_TYPE.Front);
ImageConverter converter = new ImageConverter();
newPicture.PictureData = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
```

with

```csharp
ImageConverter converter = new ImageConverter();
byte[] data = (byte[])converter.ConvertTo(picAlbumArt.Image, typeof(byte[]));
PictureInfo newPicture = PictureInfo.fromBinaryData(data, PictureInfo.PIC_TYPE.Front);
```

If Windows covers still do not work, I'd like you to try the following workarounds :
- Set `Settings.ID3v2_tagSubVersion= 3` before saving the file (writes ID3v2.3 instead of ID3v2.4, which is more Windows-friendly in general)
- Set `Settings.ID3v2_forceUnsynchronization = true` before saving the file (it seems to be MP3tag's default behaviour)

Please keep me informed !

**Comment by ElmoFuntz** [comment]
Setting the tag version to 2.3 seem to have resolved it. The other setting was not required. It does not work (consistently) if they are not set. This is odd since mp3tag is writing 2.4 and they show. I did try just ID3v2_forceUnsynchronization = true as well and it made no difference.

**Comment by Zeugma440** [comment]
Good news is we've found a way to make it work !

I'll take some time later to examine what mp3tag does and I don't.
